File: leetcode_cpp/reverse_linked_list_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        // case in which no swap is needed
        if (m == n) {
            return head;
        }

        ListNode *current = head;
        ListNode *next;
        ListNode *previous;
        int counter = 1;
        ListNode *node_m_minus_1;
        ListNode *node_m;
        ListNode *node_n;
        ListNode *node_n_plus_1;


        // case in which swap starts from the beginning
        while (true) {
            if (current != NULL) {
                next = current -> next;
            }

            if (counter == m - 1) {
                node_m_minus_1 = current;
            }
            else if (counter == m) {
                node_m = current;
            }
            else if (counter == n) {
                node_n = current;
            }
            else if (counter == n + 1) {
                node_n_plus_1 = current;
            }

            if (counter >= m + 1 && counter <= n) {
                current -> next = previous;
            }

            if (current == NULL) {
                break;
            }

            counter += 1;
            previous = current;
            current = next;
        }

        if (m != 1) {
            node_m_minus_1 -> next = node_n;
        }
        else {
            head = node_n;
        }
        node_m -> next = node_n_plus_1;

        return head;
    }
};

File: leetcode_python/reverse_linked_list_II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween(self, head, m, n):

        if m == n:
            return head

        counter = 1
        current = head

        while True:
            if not current is None:
                next = current.next

            if counter == m - 1:
                node_m_minus_1 = current
            elif counter == m:
                node_m = current
            elif counter == n:
                node_n = current
            elif counter == n + 1:
                node_n_plus_1 = current

            if (m + 1 <= counter <= n):
                current.next = previous

            if current is None:
                break

            counter += 1
            previous = current
            current = next

        if m != 1:
            node_m_minus_1.next = node_n
        else:
            head = node_n

        node_m.next = node_n_plus_1

        return head
